### ISSUE TITLE
Truncate large payloads to ensure we don't drop them

### ISSUE THREAD

**Issue by rivkahstandig3636** [rollbar/pyrollbar#237]

**Comment by DZPM** [comment]
Still happening.

Our work-around is to check the size `request.body`. If greatear than 1 MB, we drop the traceback.

**Comment by AwaishK** [comment]
@DZPM while reporting error on rollbar, I am also getting below error. errors on rollbar are being reported at different places in different files, do you know how can I trace back where this is coming from? And how to stop this error to pop up again? 

 `Failsafe from pyrollbar: payload too large. Original payload may be found in your server logs by searching for the UUID.`

Below is how I report errors on rollbar, where do you check the request.body size?
`rollbar.report_exc_info()`

**Comment by DZPM** [comment]
@AwaishK I'm afraid don't have access to that codebase anymore.
I guess you should check the size before reporting to rollbar. Good luck!

**Comment by camden-lowrance** [comment]
Any workaround for this?  I am getting this regularly when sqlalchemy throws an exception.

**Comment by bsmedberg-xometry** [comment]
Xometry is also experiencing this problem: the data that is too large is locals data from the stack frames of the exception: it would be perfectly acceptable, if locals data is too large, to partially or complete omit it from the payload.

**Comment by brianr** [comment]
Here is a blunt instrument as a workaround... you can disable locals collection entirely with:

```
rollbar.SETTINGS['locals']['enabled'] = False
```

Agree this should be handled better; @bsmedberg-xometry 's suggestion makes sense to me.

**Comment by saiyalamarty** [comment]
We had a similar issue where the default rollbar [shortener transform](https://github.com/rollbar/pyrollbar/blob/master/rollbar/lib/transforms/shortener.py) wasn't shortening parts of our payload.
We added a custom payload handler using [Link](https://docs.rollbar.com/docs/python#transforming-the-payload) to shorten the payload with custom logic.

**Comment by schmave** [comment]
Adding this to my Rollbar config solved the problem for us:

```python
        'shortener_keys': [
            ('request', 'POST', '*'),
            ('request', 'POST', '*', '*'),
            ('request', 'POST', '*', '*', '*'),
            ('body', 'trace', 'frames', '*', 'locals', '*', '*'),
            ('body', 'trace', 'frames', '*', 'locals', '*', '*', '*'),
        ]
```

The default shortener keys [defined here](https://github.com/cmu-cs-academy/pyrollbar/blob/master/rollbar/__init__.py#L417) don't truncate values inside local variable dictionaries, and they don't truncate POST values at all.

**Comment by AdrianB-sovo** [comment]
I didn't check how it could be done, but maybe Pyrollbar could have a fallback in case the configuration is not good enough?
The algorithm could look like that:
- If payload > payload_limit:
  - Get a list of representations that would be sent in the payload, ordered by their size in the payload.
  - While the size of the payload would be too large: remove the item from the payload.

I think I'd prefer data where things would be removed like that than an event without any exploitable info.

**Comment by danielmorell** [comment]
We had a bug in how we applied the shortener to deep objects. This has been fixed in `v1.1.0-alpha`.
